posts/2024-05-06-jakarta-nosql-in-action-meet-jnopo-game.adoc

```diff
@@ -0,0 +1,799 @@
+
+Firstly, `Playoffs` needs to provide a method to retrieve the non-tied game results:
+
+[source, java]
+----
+package org.jakartaee.sample.model;
+
+import jakarta.data.repository.DataRepository;
+import jakarta.data.repository.Query;
+import jakarta.data.repository.Repository;
+import jakarta.data.repository.Save;
+
+import java.util.stream.Stream;
+
+@Repository
+public interface Playoffs extends DataRepository<GameMatch,String> {
+
+    @Save
+    GameMatch add(GameMatch gameMatch);
+
+    @Query("select * from GameMatch where tied=false")
+    Stream<GameMatch> nonTiedGameMatches();
+
+}
+----
+
+NOTE: *JNoSQL* offers a simple query language pretty similar to SQL but be sure that it's not a complete SQL implementation, which means, some aggregation and functions will be not available.
+
+Regarding the aggregation process that evolves the ranking logic creation, we use to see developers delegating this aggregation logic to the DBMS and, depending on the persistence mechanism, it should be the best approach. By now, the grouping projection functions like we have with Jakarta Persistence is not available on the Jakarta NoSQL for now, but, in the future, such capability could be available as well.
+
+In order to keep our implementation free of vendor lock-in, I decided to implement the aggregation process by using the Stream API. I added comments on the code to help you to understand the implemented logic.
+
+[source, java]
+----
+package org.jakartaee.sample.model;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public record Ranking(Map<String, Integer> data) {
+    public static Ranking winnerRanking(Playoffs playoffs) {
+
+        var data =
+                // getting the game matches that are not tied
+                playoffs.nonTiedGameMatches()
+                // grouping by winner's name and summarize by game match
+                .collect(Collectors.groupingBy(
+                        g -> g.winner().name(),
+                        Collectors.collectingAndThen(Collectors.toList(), Collection::size)))
+                .entrySet()
+                .stream()
+                // sorting the results by number of game match in descending order
+                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
+                // collecting as a LinkedHashMap to keep the sorted items
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (e1, e2) -> e1, LinkedHashMap::new));
+
+        return new Ranking(data);
+    }
+}
+----
+
+Now, let make this ranking be accessible by our model. As the ranking needs the data from `Playoffs` component, why not to put this ranking creation on the `Playoffs` interface? That's we're going to do! It is possible since the Java 8:
```

Comment:
```suggestion
Now, let make this ranking be accessible by our model. As the ranking needs the data from `Playoffs` component, why not to put this ranking creation on the `Playoffs` interface? That's we're going to do! It is possible since the Java 8:
```
```suggestion
Now, let make this ranking accessible to our model. As the ranking needs the data from the `Playoffs` component, why not to put the ranking creation on the `Playoffs` interface? That's we're going to do! It is possible since Java 8:
```